main.go

```diff
@@ -61,34 +65,109 @@ func printVersion() {
 	setupLog.Info(fmt.Sprintf("Component: %s", version.String))
 }
 
-func setupChecks(mgr ctrl.Manager) {
+func setupChecks(mgr ctrl.Manager) error {
 	if err := mgr.AddReadyzCheck("ping", healthz.Ping); err != nil {
 		setupLog.Error(err, "unable to create ready check")
-		os.Exit(1)
+		return err
 	}
 
 	if err := mgr.AddHealthzCheck("ping", healthz.Ping); err != nil {
 		setupLog.Error(err, "unable to create health check")
-		os.Exit(1)
+		return err
 	}
+	return nil
+}
+
+func runController(watchNamespace string, imageServer imagehandler.ImageHandler) error {
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:                scheme,
+		Port:                  0, // Add flag with default of 9443 when adding webhooks
+		Namespace:             watchNamespace,
+		ClientDisableCacheFor: []client.Object{&corev1.Secret{}},
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start manager")
+		return err
+	}
+
+	imgReconciler := metal3iocontroller.PreprovisioningImageReconciler{
+		Client:       mgr.GetClient(),
+		Log:          ctrl.Log.WithName("controllers").WithName("PreprovisioningImage"),
+		APIReader:    mgr.GetAPIReader(),
+		Scheme:       mgr.GetScheme(),
+		ImageHandler: imageServer,
+	}
+	if err = (&imgReconciler).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "PreprovisioningImage")
+		return err
+	}
+
+	// +kubebuilder:scaffold:builder
+
+	if err := setupChecks(mgr); err != nil {
+		return err
+	}
+
+	setupLog.Info("starting manager")
+	return mgr.Start(ctrl.SetupSignalHandler())
+}
+
+func loadStaticNMState(nmstateDir string, imageServer imagehandler.ImageHandler) error {
+	files, err := ioutil.ReadDir(nmstateDir)
+	if err != nil {
+		return errors.WithMessagef(err, "problem reading %s", nmstateDir)
+	}
+
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		b, err := os.ReadFile(path.Join(nmstateDir, f.Name()))
+		if err != nil {
+			return errors.WithMessagef(err, "problem reading %s", path.Join(nmstateDir, f.Name()))
+		}
+		igBuilder := ignition.New(b,
+			os.Getenv("IRONIC_BASE_URL"),
+			os.Getenv("IRONIC_AGENT_IMAGE"),
+			os.Getenv("IRONIC_AGENT_PULL_SECRET"),
+			os.Getenv("IRONIC_RAMDISK_SSH_KEY"),
+		)
+		ign, err := igBuilder.Generate()
+		if err != nil {
+			return errors.WithMessagef(err, "problem generating ignition %s", f.Name())
+		}
+		imageName := strings.Replace(f.Name(), ".yaml", ".qcow", 1) // master-1.yaml -> master-1.qcow
```

Comment:
The assumption here it's that the filename will match the original server name

Reply:
well we don't have the server name, so this is the only thing I can use for an image name

Reply:
If we want to make it slightly more robust, we could evaluate the option to load a json file containing just two fields, the server name and the nmstate content (it could be not so difficult to generate it from within the installer).

Reply:
Yeah that's what we've been working on Andrea.  Too bad you're in a different timezone or we could work together on it.